Test: check zephyr os tracing functions.

youhuazhu opened the pull request:
Use self-define backends to get tracing data and check
the output data.


Signed-off-by: YouhuaX Zhu <youhuax.zhu@intel.com>

nashif commented on tests/subsys/tracing/prj.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_ZTEST=y
```
what are we testing here exactly?

youhuazhu replied:
Thanks @nashif ,  This is for the case when tracing functions disabled,   if tracing functions are disabled,  we shouldn't find any tracing data in the backend:
```
//checked in main.c
#ifdef CONFIG_TRACING
     zassert_true(thread_switched_found, "doesn't find thread_switched.");
     zassert_true(sys_trace_void_found, "doesn't find sys_trace_void.");
     zassert_true(sys_trace_semaphore_found,
             "doesn't find thread_idle_entered.");
#else
     /* Should produce any tracing data when tracing disabled. */
     zassert_false(thread_switched_found, "shouldn't have trace data.");
     zassert_false(sys_trace_void_found, "shouldn't have trace data.");
     zassert_false(sys_trace_semaphore_found, "shouldn't have trace data.");
#endif
```

nashif commented on tests/subsys/tracing/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+
+cmake_minimum_required(VERSION 3.13.1)
+
+if(BOARD MATCHES "qemu_.*")
```
where is this output being used?

youhuazhu replied:
not used yet, I will remove this.  thanks .

youhuazhu replied:
Have fixed this, thanks.

nashif commented on tests/subsys/tracing/prj_uart_ctf.conf:
```diff
@@ -0,0 +1,8 @@
+CONFIG_ZTEST=y
```
I do not see this file being used anywhere

youhuazhu replied:
yes, I will submit another PR for checking tracing ctf,  so will remove this file firstly.  thanks for your reminding.

youhuazhu replied:
Have removed prj_uart_ctf.conf , thanks.

nashif commented on tests/subsys/tracing/testcase.yaml:
```diff
@@ -0,0 +1,12 @@
+tests:
+  tracing.transport:
+    tags: tracing_disabled
```
what is the purpose of this? What are you testing?

youhuazhu replied:
as above reply, This is for the case when tracing functions disabled, if tracing functions are disabled, we shouldn't find any tracing data in the backend.

nashif commented on tests/subsys/tracing/testcase.yaml:
```diff
@@ -0,0 +1,12 @@
+  tracing.transport.usb:
+    extra_args: CONF_FILE="prj_usb.conf"
+    filter: CONFIG_UART_CONSOLE and not CONFIG_QEMU_TARGET
+    platform_exclude: adafruit_feather_m0_basic_proto
```
why is this being excluded, add a comment

nashif commented on tests/subsys/tracing/testcase.yaml:
```diff
@@ -0,0 +1,12 @@
+    tags: tracing_uart
+  tracing.transport.usb:
+    extra_args: CONF_FILE="prj_usb.conf"
+    filter: CONFIG_UART_CONSOLE and not CONFIG_QEMU_TARGET
```
why is qemu being excluded here? when does this test get run? I see it now only building for all tests where tracing is enabled and only runs on tests that do not enable tracing, so what is the point?

youhuazhu replied:
this tests worked for phisical boards,  and Let me have a double check if qemu support tracing output....

youhuazhu replied:
I have fixed this and enabled the qemu platform.

youhuazhu commented:
Hi @nashif ,  Have addressed all your comments,   Any ohter comment or suggestions?  thanks very much.

youhuazhu commented:
Hi @nashif , would you take a time to review this PR again? thanks.

LeiW000 commented on tests/subsys/tracing/src/main.c:
```diff
@@ -0,0 +1,145 @@
+		const struct tracing_backend *backend,
+		uint8_t *data, uint32_t length)
+{
+	uint8_t recv_char[100], recv_index = 0;
```
I don't think you have to define a new buffer. Only using "*data" is enough.

youhuazhu replied:
recv_char[] is used to store a whole line string,  and *data just get a char from backend,  so we need define a new buffer here.

youhuazhu replied:
Let me have a try,

LeiW000 commented on tests/subsys/tracing/src/main.c:
```diff
@@ -0,0 +1,145 @@
+#endif
+	generate_more_tracing_data();
+#ifdef CONFIG_TRACING
+	zassert_true(thread_switched_found, "doesn't find thread_switched.");
```
Its better to change the error message like below.
"thread_switched can't be found".

Same to line 123, 125, 129 and 130.

youhuazhu replied:
yes,  it's more clear,  I have fixed them, thanks.

youhuazhu commented on tests/subsys/tracing/src/tracing_string_format_test.c:
```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
```
done, thanks steven.

LeiW000 reviewed (approved):
It looks pefect to me now. 👍

youhuazhu commented:
Thanks @LeiW000 for review,  
Hi @nashif , Can you help to review this PR again?

youhuazhu commented:
Hi @nashif , Any other comments else?  Please help to review it,

youhuazhu commented:
Hi @nashif,  Any other coomments for this PR? since this PR waiting for a long time.

youhuazhu commented:
reabsed to the latest code base.  @nashif  please help to review.

LeiW000 commented on tests/subsys/tracing/test_ctf.py:
```diff
@@ -0,0 +1,44 @@
+        os.mkdir(data_path + "/ctf")
+
+    if os.path.exists("subsys/tracing/ctf/tsdl/metadata"):
+        os.system('cp subsys/tracing/ctf/tsdl/metadata ${data_path}/ctf/')
```
It better to use "High-level file operations" to copy file.

from shutil import copyfile
copyfile(src, dst)

youhuazhu replied:
Thanks Steven, Have changed  to copyfile.

LeiW000 commented on scripts/sanity_chk/harness.py:
```diff
@@ -121,6 +127,29 @@ def handle(self, line):
+                    stdout, _ = proc.communicate(15)
+                    if "failed" not in stdout.decode():
+                        self.state = "passed"
+                    elif "skip" in stdout.decode():
```
Please double confirm if "skip" can be shown with pytest.

youhuazhu replied:
Never show "skip" with pytest,  but test_ctf.py will show 'skip'  when babeltrace not installed.

youhuazhu committed:
Test: check zephyr os tracing functions.

Define an uart and an usb backends to get tracing output data,
then verify the content of the output data.

Signed-off-by: YouhuaX Zhu <youhuax.zhu@intel.com>

LeiW000 commented:
Ping @nashif to review it.

LeiW000 commented:
@nashif , can it be merged now?

youhuazhu commented:
Hi @nashif  Can you help to review this PR which handle tracing non-ctf case ?

Zhaoningx commented:
This PR has duplicated by https://github.com/zephyrproject-rtos/zephyr/pull/31387 , because the PR's owner Youhua resigned.
